data_migration.go

```diff
@@ -0,0 +1,53 @@
+package influxdb
+
+import (
+	"context"
+	"time"
+)
+
+// DataMigrationService will transfer the old stored data to new released version.
+type DataMigrationService interface {
+	// CheckAndMigrate will determine if data already migrated.
+	// Usually we will put an index to make sure this the newest version.
+	// Then do a scan to the storage and convert every thing related.
+	CheckAndMigrate(ctx context.Context) error
+	BucketMigrationService
+}
+
+// BucketIsMigratedIndex will be the index check to determine
+// if the newest bucket schema is applied.
+var BucketIsMigratedIndex = []byte("bucketIsMigrated_org")
+
+// BucketMigrationService will migrate old bucket to the most recent bucket schema.
+type BucketMigrationService interface {
+	IsBucketMigrated(ctx context.Context) bool
+	ConvertBucketToNew(ctx context.Context) error
+}
+
+// OldBucket should includes all old fields of previous bucket schemas,
+// as well as the new fields incase of any partial conversion.
+type OldBucket struct {
```

Comment:
What happens when we have to change the schema of buckets again?

Reply:
@desa, you can simple replace the BucketIsMigratedIndex with a new value, for example: bucketIsMigrated_orgID. any deleted value, or new added value has be show up in the oldBucket.